man/utm2ll.1
.TH UTM2LL  1

.SH NAME
utm2ll \- Convert UTM coordinates to Latitude and Longitude.


.SH SYNOPSIS
.B utm2ll 
.I zone easting northing 
.RS
.P
\fIzone\fR is UTM zone 1 thru 60 with optional latitudinal band.
.P
\fIeasting\fR is x coordinate in meters.
.P
\fInorthing\fR is y coordinate in meters.
.RE

.SH DESCRIPTION
\fBll2utm\fR  converts UTM coordinates to latitude and longitude.


.SH OPTIONS
.TP
None.


.SH EXAMPLES
.P
.B utm2ll 19T 306130 4726010
.P
latitude = 42.662139, longitude = -71.365553
.P


.SH SEE ALSO
More detailed information is in the pdf files in /usr/local/share/doc/direwolf, or possibly /usr/share/doc/direwolf, depending on installation location.

Applications in this package: aclients, atest, decode_aprs, direwolf, gen_packets, kissutil, ll2utm, log2gpx, text2tt, tt2text, utm2ll



CHANGES.md

# Revision History #


## Version 1.6  --  Under Development ##


### Bugs Fixed: ###

- Proper counting of frames in transmit queue for AGW protocol 'Y' command.



### New Features: ###


- "-t" option now accepts more values to accommodate inconsistent handling of text color control codes by different terminal emulators.  The default, 1, should work with most modern terminal types.  If the colors are not right, try "-t 9" to see the result of the different choices and pick the best one.  If none of them look right, file a bug report and specify: operating system version (e.g. Raspbian Buster), terminal emulator type and version (e.g.  LXTerminal 0.3.2).   Include a screen capture.


- "-g" option to force G3RUH mode for lower speeds where a different modem type may be the default.

- 2400 bps compatibility with MFJ-2400.  See ***2400-4800-PSK-for-APRS-Packet-Radio.pdf*** for details

- "atest -h" will display the frame in hexadecimal for closer inspection.

- Add support for Multi-GNSS NMEA sentences.



### New Documentation: ###

- ***AX.25 Throughput: Why is 9600 bps Packet Radio only twice as fast as 1200?***

- Power Point slide show in separate repository.  [https://github.com/wb2osz/direwolf-presentation ](https://github.com/wb2osz/direwolf-presentation)



## Version 1.5  --  September 2018 ##


### New Features: ###

- PTT using GPIO pin of CM108/CM119 (e.g. DMK URI, RB-USB RIM), Linux only.

- More efficient error recovery for AX.25 connected mode.  Better generation and processing of REJ and SREJ to reduce unnecessary duplicate "**I**" frames.

- New configuration option, "**V20**", for listing stations known to not understand AX.25 v2.2.  This will speed up connection by going right to SABM and not trying SABME first and failing.

- New "**NOXID**" configuration file option to avoid sending XID command to listed station(s).  If other end is a partial v2.2 implementation, which recognizes SABME, but not XID, we would waste a lot of time resending XID many times before giving up.   This is less drastic than the "**V20**" option which doesn't even attempt to use v2.2 with listed station(s).

- New application "**kissutil**" for troubleshooting a KISS TNC or interfacing to an application via files.

- KISS "Set Hardware" command to report transmit queue length.

- TCP KISS can now handle multiple concurrent applications.

- Linux can use serial port for KISS in addition to a pseudo terminal.

- decode_aprs utility can now accept KISS frames and AX.25 frames as series of two digit hexadecimal numbers.

- Full Duplex operation.  (Put "FULLDUP ON" in channel section of configuration file.)

- Time slots for beaconing.

- Allow single log file with fixed name rather than starting a new one each day.



### Bugs Fixed: ###

- Possible crash when CDIGIPEAT did not include the optional alias.

- PACLEN configuration item no longer restricts length of received frames.

- Strange failures when trying to use multiple KISS client applications over TCP.  Only Linux was affected.  

- Under certain conditions, outgoing connected mode data would get stuck in a queue and not be transmitted.  This could happen if client application sends a burst of data larger than the "window" size (MAXFRAME or EMAXFRAME option).


- Little typographical / spelling errors in messages.


### Documentation: ###


- New document ***Bluetooth-KISS-TNC.pdf*** explaining how to use KISS over Bluetooth.

- Updates describing cheap SDR frequency inaccuracy and how to compensate for it.

### Notes: ###

Windows binary distribution now uses gcc (MinGW) version 6.3.0.

----------

## Version 1.4  -- April 2017 ##


### New Features: ###

- AX.25 v2.2 connected mode.  See chapter 10 of User Guide for details.

- New client side packet filter to select "messages" only to stations that have been heard nearby recently.  This is now the default if no IS to RF filter is specified.

- New beacon type, IBEACON, for sending IGate statistics.

- Expanded debug options so you can understand what is going on with packet filtering.

- Added new document ***Successful-APRS-IGate-Operation.pdf*** with IGate background, configuration, and troubleshooting tips.
- 2400 & 4800 bps PSK modems.  See ***2400-4800-PSK-for-APRS-Packet-Radio.pdf*** in the doc directory for discussion.

- The top speed of 9600 bps has been increased to 38400.  You will need a sound card capable of 96k or 192k samples per second for the higher rates.  Radios must also have adequate bandwidth.  See ***Going-beyond-9600-baud.pdf*** in the doc directory for more details.

- Better decoder performance for 9600 and higher especially for low audio sample rate to baud ratios.

- Generate waypoint sentences for use by AvMap G5 / G6 or other mapping devices or applications.   Formats include
 - $GPWPL	- NMEA generic with only location and name.
 - $PGRMW	- Garmin, adds altitude, symbol, and comment to previously named waypoint.
 - $PMGNWPL	- Magellan, more complete for stationary objects.
 - $PKWDWPL	- Kenwood with APRS style symbol but missing comment.


- DTMF tones can be sent by putting "DTMF" in the destination address,  similar to the way that Morse Code is sent.

- Take advantage of new 'gpio' group and new /sys/class/gpio ownership in Raspbian Jessie.

- Handle more complicated gpio naming for CubieBoard, etc.

- More flexible dw-start.sh start up script for both GUI and CLI environments.


 
### Bugs Fixed: ###

- The transmitter (PTT control) was being turned off too soon when sending Morse Code.

- The -qd (quiet decode) command line option now suppresses errors about improperly formed Telemetry packets.

- Longer tocall.txt files can now be handled.  

- Sometimes kissattach would have an issue with the Dire Wolf pseudo terminal.  This showed up most often on Raspbian but sometimes occurred with other versions of Linux.

	*kissattach: Error setting line discipline: TIOCSETD: Device or resource busy
	Are you sure you have enabled MKISS support in the kernel

utm2ll.c
/* UTM to Latitude / Longitude conversion */

#include "direwolf.h"

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <ctype.h>

#include "utm.h"
#include "mgrs.h"
#include "usng.h"
#include "error_string.h"


#define D2R(d) ((d) * M_PI / 180.)
#define R2D(r) ((r) * 180. / M_PI)




static void usage();


int main (int argc, char *argv[]) 
{
	double easting;
	double northing;
	double lat, lon;
	char szone[100];
	long lzone;
	char *zlet;
	char hemi;
	long err;
	char message[300];


	if (argc == 4) {

// 3 command line arguments for UTM

	  strlcpy (szone, argv[1], sizeof(szone));
	  lzone = strtoul(szone, &zlet, 10);

	  if (*zlet == '\0') {
	    hemi = 'N';
	  }
	  else {
	  
	    if (islower(*zlet)) {
	      *zlet = toupper(*zlet);
	    }
	    if (strchr ("CDEFGHJKLMNPQRSTUVWX", *zlet) == NULL) {
	      fprintf (stderr, "Latitudinal band must be one of CDEFGHJKLMNPQRSTUVWX.\n\n");
	      usage();
	    }
	    if (*zlet >= 'N') {
	      hemi = 'N';
	    }
	    else {
	      hemi = 'S';
	    }
	  }
  
	  easting = atof(argv[2]);

	  northing = atof(argv[3]);

	  err = Convert_UTM_To_Geodetic(lzone, hemi, easting, northing, &lat, &lon);
	  if (err == 0) {
	    lat = R2D(lat);
	    lon = R2D(lon);

	    printf ("from UTM, latitude = %.6f, longitude = %.6f\n", lat, lon);
	  }
	  else {

	    utm_error_string (err, message);
	    fprintf (stderr, "Conversion from UTM failed:\n%s\n\n", message);

	  }
	}
	else if (argc == 2) {

// One command line argument, USNG or MGRS.

// TODO: continue here.


	  err = Convert_USNG_To_Geodetic (argv[1], &lat, &lon);
 	  if (err == 0) {
	    lat = R2D(lat);
	    lon = R2D(lon);
	    printf ("from USNG, latitude = %.6f, longitude = %.6f\n", lat, lon);
	  }
	  else {
	    usng_error_string (err, message);
	    fprintf (stderr, "Conversion from USNG failed:\n%s\n\n", message);
	  }

	  err = Convert_MGRS_To_Geodetic (argv[1], &lat, &lon);
 	  if (err == 0) {
	    lat = R2D(lat);
	    lon = R2D(lon);
	    printf ("from MGRS, latitude = %.6f, longitude = %.6f\n", lat, lon);
	  }
	  else {
	    mgrs_error_string (err, message);
	    fprintf (stderr, "Conversion from MGRS failed:\n%s\n\n", message);
	  }

	}
	else {
	  usage();
	}

	exit (0);
}


static void usage (void)
{
	fprintf (stderr, "UTM to Latitude / Longitude conversion\n");
	fprintf (stderr, "\n");
	fprintf (stderr, "Usage:\n");
	fprintf (stderr, "\tutm2ll  zone  easting  northing\n");
	fprintf (stderr, "\n");
	fprintf (stderr, "where,\n");
	fprintf (stderr, "\tzone is UTM zone 1 thru 60 with optional latitudinal band.\n");
	fprintf (stderr, "\teasting is x coordinate in meters\n");
	fprintf (stderr, "\tnorthing is y coordinate in meters\n");
	fprintf (stderr, "\n");
	fprintf (stderr, "or:\n");
	fprintf (stderr, "\tutm2ll  x\n");
	fprintf (stderr, "\n");
	fprintf (stderr, "where,\n");
	fprintf (stderr, "\tx is USNG or MGRS location.\n");
	fprintf (stderr, "\n");
	fprintf (stderr, "Examples:\n");
	fprintf (stderr, "\tutm2ll 19T 306130 4726010\n");
	fprintf (stderr, "\tutm2ll 19TCH06132600\n");

	exit (1);
}